File: python/969. Pancake Sorting.py
class Solution(object):
    def pancakeSort(self, A):
        """
        :type A: List[int]
        :rtype: List[int]
        """
        ans = []
        n = len(A)
        for i in range(n, 0, -1):
            idx = A.index(i)
            if idx == i-1:
                continue
            ans.append(idx + 1)
            A = A[:idx+1][::-1] + A[idx+1:] # 反转[0, idx]之间的元素，将A[idx]这个最大值放到数组头
            A = A[:i][::-1] + A[i:] # 反转[0:i-1], 将A[0]放到A[i-1], 最大值在数组尾
            ans.append(i)
            print A
        return ans

File: go/src/leetcode/969/969. Pancake Sorting.go
/**
@author ZhengHao Lou
Date    2022/02/19
*/
package main

import "fmt"

func pancakeSort(arr []int) []int {
    var ans []int
    n := len(arr)
    for l := n; l > 0; l-- {
        var mx, idx int = 0, -1
        for i := 0; i < l; i++ {
            if arr[i] > mx {
                mx = arr[i]
                idx = i
            }
        }
        if idx == l-1 {
            continue
        }
        reverse(arr, 0, idx)
        ans = append(ans, idx+1)
        reverse(arr, 0, l-1)
        ans = append(ans, l)
    }

    fmt.Println(ans, arr)
    return ans
}

func reverse(arr []int, l, r int) {
    for l < r {
        arr[l], arr[r] = arr[r], arr[l]
        l++
        r--
    }
}

func main() {
    pancakeSort([]int{3, 2, 4, 1})
    pancakeSort([]int{1, 2, 3})
}
